Tries improvements

gregw opened the pull request:
WIP on better selection of trie based on alphabet
Also some ideas on how to improve impls in TODOs

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Tries improvements

WIP on better selection of trie based on alphabet
Also some ideas on how to improve impls in TODOs

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
WIP

+ CaseSensitive ArrayTrie
+ Better space efficiency in big lookups
+ Better test cases

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTrie.java:
```diff
@@ -56,7 +56,8 @@
      * characters.
      */
-    private static final int ROW_SIZE = 32;
+    private static final int ROW_SIZE = 48;
```
It's certainly worth measuring the space & time difference between row sizes of 32 and 48.

gregw replied:
My gut feeling is that the extra space for digits means that we can now more often just allocate the normal table and avoid the secondary larger table more often.

gregw committed:
Merge remote-tracking branch 'origin/jetty-10.0.x' into jetty-10.0.x-trie-selection

gregw committed:
WIP

Added test for empty key

gregw committed:
WIP

Added more tests for exceptional values

gregw committed:
Issue #5304 HTTP2 Host Header

refactored common code
fast alphabet lookup

gregw committed:
WIP

big cleanup of code to reduce parallel array lookups

gregw committed:
WIP

tail recursion fix for required node calculation

gregw committed:
WIP

more tests

Signed-off-by: Greg Wilkins <gregw@webtide.com>

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/Index.java:
```diff
@@ -193,42 +226,55 @@
              */
             public Mutable<V> build()
             {
-                if (contents != null && maxCapacity == 0)
```
Why did you get rid of this check? IMHO building an index both pre-loaded with some contents and a max capacity of zero looks like a programming mistake we want to catch.

gregw replied:
Then we need another check as returning an empty Trie wont catch programming errors

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/Index.java:
```diff
@@ -193,42 +226,55 @@
-                {
-                    result = EmptyTrie.instance(caseSensitive);
+                    if (contents != null)
+                        throw new IllegalStateException("Cannot create a mutable index without alphabet or some contents");
```
Should we define a default alphabet instead of throwing here?

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/Index.java:
```diff
@@ -19,18 +19,26 @@
  */
 public interface Index<V>
 {
+    Set<Character> VISIBLE_ASCII_ALPHABET = Collections.unmodifiableSet(IntStream.range(0x20, 0x7f)
```
I do not like this constant very much, I would prefer an extra helper method on the builder:

```java
 public Builder<V> withVisibleAsciiAlphabet()
 {
  alphabet = Collections.unmodifiableSet(...);
  return this;
 }
```

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/Index.java:
```diff
@@ -354,11 +401,26 @@ public Builder()
+            if (trie != null)
+                return trie;
+
+            throw new UnsupportedOperationException("No suitable Trie implementation : " + this);
```
You should throw `IllegalStateException` here, like the mutable builder does.

gregw committed:
WIP transfer machines

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
WIP

TernaryTrie mostly working

gregw committed:
WIP

growby of TernaryTrie

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw commented:
@lorban Can you run the new trie through your trie benchmark

gregw committed:
WIP

update from review

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
WIP

update javadoc

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Merge remote-tracking branch 'origin/jetty-10.0.x' into jetty-10.0.x-trie-selection

gregw committed:
WIP

Fixed some issues with alphabets

gregw committed:
WIP

Fixed null value tests

gregw committed:
WIP

Don't use an Index for HTTP2Cipher

gregw committed:
WIP

Obeying the orders of my IDE

gregw committed:
WIP

removed test main

joakime commented on jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Cipher.java:
```diff
@@ -19,311 +19,310 @@
+        "SSL",
+        "SSLv2",
+        "SSLv3"
+    ).map(String::toUpperCase).collect(Collectors.toSet());
```
This `String::toUpperCase` uses system default locale.
This will produce undesired side effects on on some, non-US, systems.

joakime commented on jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Cipher.java:
```diff
@@ -19,311 +19,310 @@
+        "TLS_PSK_WITH_AES_256_CCM",
+        "TLS_PSK_WITH_AES_128_CCM_8",
+        "TLS_PSK_WITH_AES_256_CCM_8"
+    ).map(String::toUpperCase).collect(Collectors.toSet());
```
This `String::toUpperCase` uses system default locale.
This will produce undesired side effects on on some, non-US, systems.

joakime commented on jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Cipher.java:
```diff
@@ -19,311 +19,310 @@
     public static boolean isBlackListProtocol(String tlsProtocol)
     {
-        return __blackProtocols.get(tlsProtocol) != null;
+        return __blackProtocols.contains(tlsProtocol.toUpperCase());
```
This uses system default locale.
Change to `tlsProtocol.toUpperCase(Locale.US)`.
This will produce undesired side effects on on some, non-US, systems.

joakime commented on jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Cipher.java:
```diff
@@ -19,311 +19,310 @@
     public static boolean isBlackListCipher(String tlsCipher)
     {
-        return __blackCiphers.get(tlsCipher) != null;
+        return __blackCiphers.contains(tlsCipher.toUpperCase());
```
This uses system default locale.
Change to `tlsProtocol.toUpperCase(Locale.US)`.
This will produce undesired side effects on on some, non-US, systems.

joakime reviewed (changes requested):
default `Locale` issues spotted

gregw committed:
WIP

removed locale issue with case conversion.

lorban commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java:
```diff
@@ -1064,18 +1063,20 @@ else if (_endOfContent == EndOfContent.CHUNKED_CONTENT)
-                            ? new Index.Builder<HttpField>()
+                            ? new Index.Mutable.Builder<HttpField>()
                             .caseSensitive(false)
                             .mutable()
```
This call isn't needed anymore.

gregw replied:
Unfortunately it is because `caseInsensitvie(false)` returns the wrong type of Builder.  I guess I could reorder to make that last.... or just not do the alternative constructor....

lorban replied:
Right... by exposing this builder's constructor you've opened a whole new range of potential maintenance problems. We have to seriously ponder if it's worth it.

gregw replied:
yep - I don't like it... it is a halfway house.  If the builder is not efficient enough, we should go directly to the impl, or have a very special purpose static method builder.... let me try those approaches....

lorban commented on jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Cipher.java:
```diff
@@ -19,311 +19,312 @@
-import org.eclipse.jetty.util.Index;
+import org.eclipse.jetty.util.StringUtil;
 
 public class HTTP2Cipher
```
Looking at the changes made to this class, two things pop into my mind:
 - Do those maps need to be case-insensitive?
 - Aren't there other classes that make use of Index that should instead use a Map?

gregw replied:
I think yes, unless we can find something that specifies that protocol and ciphernames are specific cases.

There are probably other uses of Index that could be a map... I will look...

gregw replied:
There are a couple of maybe could be changed... but nothing that shouts YES CHANGE ME NOW!.  So I'm happy as is.

lorban reviewed (changes requested):
There more I look at this PR, the more I think the tries implementations and Index.Builder improvements should be in a different PR than the "Index vs Map" changes.

gregw committed:
Merge remote-tracking branch 'origin/jetty-10.0.x' into jetty-10.0.x-trie-selection

gregw commented:
@lorban if I find other cases that should be converted to a Map, then I'll spin out a different PR.  Otherwise I leave it as a twofor PR

gregw committed:
Updates from Review

 + Remove public mutable builder constructor
 + align caseInsensitive vs caseSensitive

gregw committed:
fix javadoc

gregw committed:
wip

Added a list parser to Index so that getBest can be used rather than clumsy calls to best

gregw committed:
Merge remote-tracking branch 'origin/jetty-10.0.x' into jetty-10.0.x-trie-selection

gregw committed:
Fix header

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw commented:
@joakime @lorban nudge

gregw committed:
WIP

Added benchmark along with simple HashTrie for comparison.

gregw commented:
I've worked on the JMH benchmark a bit and the latest results are:
```
Benchmark                       (TRIE_TYPE)   Mode  Cnt        Score        Error  Units
TrieBenchmark.testGet             ArrayTrie  thrpt    3  2626126.840 ± 259215.252  ops/s
TrieBenchmark.testGet           TernaryTrie  thrpt    3  1643146.585 ± 105984.724  ops/s
TrieBenchmark.testGet      ArrayTernaryTrie  thrpt    3  2141509.426 ± 163937.703  ops/s
TrieBenchmark.testGet              TreeTrie  thrpt    3  2194125.587 ± 180112.230  ops/s
TrieBenchmark.testGet              HashTrie  thrpt    3  1712587.264 ± 255330.228  ops/s

TrieBenchmark.testGetBest         ArrayTrie  thrpt    3  2015488.908 ± 444844.971  ops/s
TrieBenchmark.testGetBest       TernaryTrie  thrpt    3   945423.852 ±  57482.449  ops/s
TrieBenchmark.testGetBest  ArrayTernaryTrie  thrpt    3  1334164.632 ±  76566.334  ops/s
TrieBenchmark.testGetBest          TreeTrie  thrpt    3  1730735.183 ± 306228.766  ops/s
TrieBenchmark.testGetBest          HashTrie  thrpt    3   682218.281 ± 139110.408  ops/s

TrieBenchmark.testPut             ArrayTrie  thrpt    3    11713.479 ±   9412.908  ops/s
TrieBenchmark.testPut           TernaryTrie  thrpt    3    15903.540 ±   4547.753  ops/s
TrieBenchmark.testPut      ArrayTernaryTrie  thrpt    3    18531.887 ±   7989.733  ops/s
TrieBenchmark.testPut              TreeTrie  thrpt    3    10317.879 ±   1072.890  ops/s
TrieBenchmark.testPut              HashTrie  thrpt    3    20786.860 ±  10014.047  ops/s
```

This shows the new `TernaryTrie` class is not very good at all, so it might not be possible to have the best of both worlds.. ie unlimited alphabet as in `ArrayTernaryTrie` and unlimited size as in `TreeTrie`.

It is interesting that `TernaryTrie` is so much slower that `ArrayTernaryTrie` as it is essentially the same algorithm, but with some key fields instead of being final for the whole Trie are only final within a referenced table.  The extra de-reference plus the need to follow table points appears to be fatal to what the JIT can do.

Also interesting is how the `HashTrie` is actually not that competitive, even for `testGet`, although it has good put peformance.

`ArrayTrie`  rules both get benchmarks, but is poor in put.... which is a pity because it is used for the dynamic HttpField cache.  So I'm going to look at how it's put performance can be improved as well.

Ultimately what we want is `ArrayTrie` get performance, with `HashTrie` put performance, with the unlimited alphabet of `ArrayTernaryTrie`  and the unlimited size of `TreeTrie`.

gregw commented:
With GC

```
Benchmark                                                        (TRIE_TYPE)   Mode  Cnt        Score        Error   Units
TrieBenchmark.testGet                                              ArrayTrie  thrpt    3  2549681.339 ± 177386.537   ops/s
TrieBenchmark.testGet:·gc.alloc.rate.norm                          ArrayTrie  thrpt    3       ≈ 10⁻⁴                 B/op
TrieBenchmark.testGet                                            TernaryTrie  thrpt    3  1629659.617 ± 199081.125   ops/s
TrieBenchmark.testGet:·gc.alloc.rate.norm                        TernaryTrie  thrpt    3       ≈ 10⁻⁴                 B/op
TrieBenchmark.testGet                                       ArrayTernaryTrie  thrpt    3  2093112.400 ± 224581.021   ops/s
TrieBenchmark.testGet:·gc.alloc.rate.norm                   ArrayTernaryTrie  thrpt    3       ≈ 10⁻⁴                 B/op
TrieBenchmark.testGet                                               TreeTrie  thrpt    3  2150444.338 ± 580224.415   ops/s
TrieBenchmark.testGet:·gc.alloc.rate.norm                           TreeTrie  thrpt    3       ≈ 10⁻⁴                 B/op
TrieBenchmark.testGet                                               HashTrie  thrpt    3  1649333.103 ± 607382.407   ops/s
TrieBenchmark.testGet:·gc.alloc.rate.norm                           HashTrie  thrpt    3     1056.075 ±      0.058    B/op

TrieBenchmark.testGetBest                                          ArrayTrie  thrpt    3  2007848.358 ± 143809.756   ops/s
TrieBenchmark.testGetBest:·gc.alloc.rate.norm                      ArrayTrie  thrpt    3       ≈ 10⁻⁴                 B/op
TrieBenchmark.testGetBest                                        TernaryTrie  thrpt    3   925738.494 ± 626022.700   ops/s
TrieBenchmark.testGetBest:·gc.alloc.rate.norm                    TernaryTrie  thrpt    3       ≈ 10⁻⁴                 B/op
TrieBenchmark.testGetBest                                   ArrayTernaryTrie  thrpt    3  1400928.640 ± 862702.497   ops/s
TrieBenchmark.testGetBest:·gc.alloc.rate.norm               ArrayTernaryTrie  thrpt    3       ≈ 10⁻⁴                 B/op
TrieBenchmark.testGetBest                                           TreeTrie  thrpt    3  1543766.020 ± 101526.996   ops/s
TrieBenchmark.testGetBest:·gc.alloc.rate.norm                       TreeTrie  thrpt    3       ≈ 10⁻⁴                 B/op
TrieBenchmark.testGetBest                                           HashTrie  thrpt    3   793271.352 ±  27724.016   ops/s
TrieBenchmark.testGetBest:·gc.alloc.rate.norm                       HashTrie  thrpt    3     1384.122 ±      0.175    B/op

TrieBenchmark.testPut                                              ArrayTrie  thrpt    3    14113.322 ±   8327.750   ops/s
TrieBenchmark.testPut:·gc.alloc.rate.norm                          ArrayTrie  thrpt    3    28010.946 ±     21.586    B/op
TrieBenchmark.testPut                                            TernaryTrie  thrpt    3    17236.118 ±   1751.296   ops/s
TrieBenchmark.testPut:·gc.alloc.rate.norm                        TernaryTrie  thrpt    3    19937.551 ±      0.165    B/op
TrieBenchmark.testPut                                       ArrayTernaryTrie  thrpt    3    18404.279 ±   3751.036   ops/s
TrieBenchmark.testPut:·gc.alloc.rate.norm                   ArrayTernaryTrie  thrpt    3    15713.452 ±      0.294    B/op
TrieBenchmark.testPut                                               TreeTrie  thrpt    3    10415.684 ±   4583.961   ops/s
TrieBenchmark.testPut:·gc.alloc.rate.norm                           TreeTrie  thrpt    3   331930.999 ±      8.612    B/op
TrieBenchmark.testPut                                               HashTrie  thrpt    3    19072.713 ±   9326.126   ops/s
TrieBenchmark.testPut:·gc.alloc.rate.norm                           HashTrie  thrpt    3    36891.538 ±     13.903    B/op
```

joakime commented:
Did you use the jmh `Blackhole` to consume the output of those tests that returned `≈ 10⁻⁴` on the GC?
If you didn't then those GC numbers are bogus, as your JIT just optimized away a whole bunch of behaviors that impact GC.

gregw committed:
WIP

reinstated TreeTrie and ArrayTernaryTrie

gregw committed:
WIP

ArrayTrie with unlimited alphabet

gregw commented:
I have made the ArrayTrie handle arbitrary alphabet and the TreeTrie able to do case sensitive, so these two tries can now do it all and are the best performance. ArrayTrie for capacity limited and TreeTrie for unlimited:
```
TrieBenchmark.testGet                                              ArrayTrie  thrpt    3  2557123.320 ± 675499.293   ops/s
TrieBenchmark.testGet:·gc.alloc.rate.norm                          ArrayTrie  thrpt    3       ≈ 10⁻⁴                 B/op
TrieBenchmark.testGet                                            TernaryTrie  thrpt    3  1648398.941 ± 285775.706   ops/s
TrieBenchmark.testGet:·gc.alloc.rate.norm                        TernaryTrie  thrpt    3       ≈ 10⁻⁴                 B/op
TrieBenchmark.testGet                                       ArrayTernaryTrie  thrpt    3  2159279.864 ± 179702.819   ops/s
TrieBenchmark.testGet:·gc.alloc.rate.norm                   ArrayTernaryTrie  thrpt    3       ≈ 10⁻⁴                 B/op
TrieBenchmark.testGet                                               TreeTrie  thrpt    3  2064166.324 ± 176390.183   ops/s
TrieBenchmark.testGet:·gc.alloc.rate.norm                           TreeTrie  thrpt    3       ≈ 10⁻⁴                 B/op
TrieBenchmark.testGet                                               HashTrie  thrpt    3  1573776.382 ±  12296.099   ops/s
TrieBenchmark.testGet:·gc.alloc.rate.norm                           HashTrie  thrpt    3     1056.074 ±      0.001    B/op

TrieBenchmark.testGetBest                                          ArrayTrie  thrpt    3  1960646.626 ± 218280.996   ops/s
TrieBenchmark.testGetBest:·gc.alloc.rate.norm                      ArrayTrie  thrpt    3       ≈ 10⁻⁴                 B/op
TrieBenchmark.testGetBest                                        TernaryTrie  thrpt    3   953142.762 ±  80495.965   ops/s
TrieBenchmark.testGetBest:·gc.alloc.rate.norm                    TernaryTrie  thrpt    3       ≈ 10⁻⁴                 B/op
TrieBenchmark.testGetBest                                   ArrayTernaryTrie  thrpt    3  1431552.933 ± 131589.264   ops/s
TrieBenchmark.testGetBest:·gc.alloc.rate.norm               ArrayTernaryTrie  thrpt    3       ≈ 10⁻⁴                 B/op
TrieBenchmark.testGetBest                                           TreeTrie  thrpt    3  1415461.535 ±  34190.999   ops/s
TrieBenchmark.testGetBest:·gc.alloc.rate.norm                       TreeTrie  thrpt    3       ≈ 10⁻⁴                 B/op
TrieBenchmark.testGetBest                                           HashTrie  thrpt    3   790983.679 ±  27521.729   ops/s
TrieBenchmark.testGetBest:·gc.alloc.rate.norm                       HashTrie  thrpt    3     1384.118 ±      0.069    B/op

TrieBenchmark.testPut                                              ArrayTrie  thrpt    3    13745.709 ±   3919.697   ops/s
TrieBenchmark.testPut:·gc.alloc.rate.norm                          ArrayTrie  thrpt    3    24010.056 ±      3.905    B/op
TrieBenchmark.testPut                                            TernaryTrie  thrpt    3    15301.310 ±   6434.233   ops/s
TrieBenchmark.testPut:·gc.alloc.rate.norm                        TernaryTrie  thrpt    3    19937.734 ±      0.822    B/op
TrieBenchmark.testPut                                       ArrayTernaryTrie  thrpt    3    17726.418 ±    769.060   ops/s
TrieBenchmark.testPut:·gc.alloc.rate.norm                   ArrayTernaryTrie  thrpt    3    15713.523 ±      6.210    B/op
TrieBenchmark.testPut                                               TreeTrie  thrpt    3    11094.200 ±    872.619   ops/s
TrieBenchmark.testPut:·gc.alloc.rate.norm                           TreeTrie  thrpt    3   331927.734 ±      5.938    B/op
TrieBenchmark.testPut                                               HashTrie  thrpt    3    20165.527 ±   3663.388   ops/s
TrieBenchmark.testPut:·gc.alloc.rate.norm                           HashTrie  thrpt    3    36891.142 ±      0.529    B/op
```

gregw committed:
WIP

gregw committed:
WIP

Fixed otherChars

gregw committed:
WIP

Removed TErnaryTrie
Include null key in capacity calculations

gregw committed:
WIP

misc cleanups

gregw committed:
Misc cleanups

found by IDE code analysis

gregw committed:
Merge branch 'jetty-10.0.x' into jetty-10.0.x-trie-selection

gregw commented:
@lorban @joakime nudge

gregw commented:
@lorban @joakime Can I get this reviewed before 2022?

After a long journey, this PR landed on a moderately simple solution.    By making `ArrayTrie` handle arbitrary alphabets and both `ArrayTrie` and `TreeTrie` are now able to do case sensitive, then the Index builder no longer needs to worry about alphabets or sensitively when picking an implementation.  It is really just a choice between fixed|large vs growing|unlimited capacity when picking the Trie implementation.

I also increased the row size in `ArrayTrie` to trade off a bit of space for speed.  This was mostly due to the case sensitive mode needing more slots to be sensible.   There is probably some tuning that could be done on the actual index table (or even dynamic index tables), but I don't think that is justified at this stage.

Ternary tries in all their incarnations are just slow, as they have to visit too many nodes going left/right.   So they are deprecated.

Multi-word characters are not supported.
Multi-byte characters are supported if they are a single character and the String API is used.

@sbordet did suggest that we could actually use utf-8 byte representation when building the tries, and then multi-byte characters would work from byte[] API as well.   But that is a cost when building tries, that will badly affect the dynamic HTTP field cache.... there is also not a use-case for it, as HTTP fields are not UTF-8.   So I did not change that either.

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTernaryTrie.java:
```diff
@@ -96,10 +97,28 @@
+            return null;
+        if (capacity < maxCapacity)
+            capacity = maxCapacity;
+        capacity++;
```
That increment should be documented.

gregw replied:
It's no longer needed.  Removing

lorban replied:
It looks like some tests checking the exact capacity calculation might be needed...

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/Index.java:
```diff
@@ -188,42 +269,42 @@
+
+                // check capacities
+                if (maxCapacity >= 0 && capacity > maxCapacity)
+                    throw new IllegalStateException("Insufficient maxCapacity for contents");
```
Is `AbstractTrie.requiredCapacity` reporting the exact amount of required bytes for each and every trie implementation? Otherwise we have a problem.

gregw replied:
It is if we assume all our Trie implementations can handle a key of "".
The other thing to do is to just not handle a "" and get rid of the whole +1 logic.... but then the nature of these beasts is that they have a root and the root can store a value, so hmmmm  keeping the +1 is probably simpler.    I'll document.

lorban replied:
`ArrayTernaryTrie` needs more than just +1, see https://github.com/eclipse/jetty.project/blob/2662be0a05e14fad80defd722bc191ba201c5e8b/jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTernaryTrie.java#L125-L128

gregw replied:
That comment is wrong, the +1 is enough.  Anyway, that class is deprecated and never directly called from the Index factory now.  I'll fix the comment.

gregw replied:
more javadoc improvements

lorban replied:
The `ArrayTernaryTrie` capacity calculation is really a pain, but since that implementation isn't used by the `Index` builders anymore, this looks good enough.

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/Index.java:
```diff
@@ -251,6 +333,21 @@ public Builder()
+        {
+            if (maxCapacity <= 0)
+                return EmptyTrie.instance(true);
+            if (maxCapacity <= ArrayTrie.MAX_CAPACITY)
```
If `maxCapacity` is greater than `ArrayTrie.MAX_CAPACITY` you're going to return an empty trie. This breaks the principle of least surprise IMHO.

gregw replied:
ah yes... a TreeTrie it is then

lorban replied:
Oh, and I just remembered that a negative `maxCapacity` means unlimited. I think the whole method might need to be revisited.

gregw replied:
revisited

lorban replied:
Looks good now.

lorban commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannelOverHttp.java:
```diff
@@ -500,31 +501,24 @@ public void parsedHeader(HttpField field)
                 case EXPECT:
                 {
-                    if (HttpVersion.HTTP_1_1.equals(_requestBuilder.version()))
+                    if (!Index.parseCsvIndex(HttpHeaderValue.CACHE, value, t ->
```
The `parseCsvIndex` method looks very specific to the `HttpChannelOverHttp`. It feels like it should be moved into this class.

gregw replied:
Moved and fixed multiple bugs.
There are some other possible usages of the method, but will do those in another PR

lorban commented on tests/jetty-jmh/pom.xml:
```diff
@@ -72,6 +72,11 @@
+    <dependency>
+      <groupId>org.openjdk.jol</groupId>
+      <artifactId>jol-core</artifactId>
+      <version>0.14</version>
```
The version should be moved up in the root pom's `dependencyManagement`. But it also looks like you're not sizing objects in the benchmark you added - is that an overlook or something you don't care about?

gregw replied:
it wasn't even needed

lorban reviewed (changes requested):
Here's my review, happy 2021!

gregw committed:
Merge branch 'jetty-10.0.x' into jetty-10.0.x-trie-selection

gregw committed:
Updates from review

gregw committed:
Updates from review

gregw committed:
Updates from review

joakime reviewed (approved):
Looks good, as-is.  Might want to add a few more testcases for the new lowercase/uppercase StringUtil methods in a future PR.
